Readme.md
Rundeck EC2 Nodes Plugin
========================

Version: 1.5

[![Build Status](https://travis-ci.org/rundeck-plugins/rundeck-ec2-nodes-plugin.svg?branch=master)](https://travis-ci.org/rundeck-plugins/rundeck-ec2-nodes-plugin)

This is a Resource Model Source plugin for [RunDeck][] 1.5+ that provides
Amazon EC2 Instances as nodes for the RunDeck server.

[RunDeck]: http://rundeck.org

NOTE: For Rundeck 1.4, you will need to use plugin [version 1.2][].

[version 1.2]: https://github.com/gschueler/rundeck-ec2-nodes-plugin/tree/v1.2

Installation
------------

Download from the [releases page](https://github.com/rundeck-plugins/rundeck-ec2-nodes-plugin/releases).

Put the `rundeck-ec2-nodes-plugin-1.5.x.jar` into your `$RDECK_BASE/libext` dir.

Release Notes
-------------

[Release Notes](https://github.com/rundeck-plugins/rundeck-ec2-nodes-plugin/blob/master/Release-Notes.md)

Usage
-----

You can configure the Resource Model Sources for a project either via the
RunDeck GUI, under the "Admin" page, or you can modify the `project.properties`
file to configure the sources.

See: [Resource Model Source Configuration](http://rundeck.org/1.5/manual/plugins.html#resource-model-source-configuration)

The provider name is: `aws-ec2`

Here are the configuration properties:

* `accessKey`: API AccessKey value (if not using IAM profile)
* `secretKey`: API SecretKey value (if not using IAM profile)
* `endpoint` - the URL of the AWS **endpoint** to use, or blank for the default endpoint (see [Amazon EC2 Regions and Endpoints](http://docs.aws.amazon.com/general/latest/gr/rande.html#ec2_region))
* `synchronousLoad`: Do not use internal async loading behavior. (boolean, default: true)
* `refreshInterval`: Unless using Synchronous Loading, time in seconds used as minimum interval between calls to the AWS API. (default 30)
* `filter` A set of ";" separated query filters ("$Name=$Value") for the AWS EC2 API, see below.
* `runningOnly`: if "true", automatically filter the * instances by "instance-state-name=running"
* `useDefaultMapping`: if "true", base all mapping definitions off the default mapping provided.
* `mappingParams`: A set of ";" separated mapping entries
* `mappingFile`: Path to a java properties-formatted mapping definition file.

If you leave `accessKey` and `secretKey` blank, the EC2 IAM profile will be used.

Note: Rundeck 2.6.3+ uses an asynchronous nodes cache by
default. You should enable `synchronousLoad` if you are using the
rundeck nodes cache, or set the `refreshInterval` to 0.

## Filter definition

The syntax for defining filters uses `$Name=$Value1,$Value2[;$Name=$value[;...]]`
for any of the allowed filter names (see [DescribeInstances][1] for the available filter Names).

*Note*: you do not need to specify `Filter.1.Name=$Name`, etc. as described in the EC2 API documentation,
this will handled for you.  Simply list the Name = Value pairs, separated by `;`.

You can specify multiple values in the filter by separating the values with `,`.

 [1]: http://docs.amazonwebservices.com/AWSEC2/latest/APIReference/ApiReference-query-DescribeInstances.html

Example: to filter based on a Tag named "MyTag" with a value of "Some Tag Value":

    tag:MyTag=Some Tag Value
    
Example: to filter *any* instance with a Tag named `MyTag`:

    tag-key=MyTag

Example combining matching a tag value and the instance type:

    tag:MyTag=Some Tag Value;instance-type=m1.small

Example including two instance types, the results will have one or the other instance-type:

    instance-type=m1.small,m1.large

Mapping Definition
----------

RunDeck Node attributes are configured by mapping EC2 Instance properties via a
mapping configuration.

The mapping declares the node attributes that will be set, and what their values
will be set to using a "selector" on properties of the EC2 Instance object.

Here is the default mapping:

    description.default=EC2 node instance
    editUrl.default=https://console.aws.amazon.com/ec2/home#s=Instances&selectInstance=${node.instanceId}
    hostname.selector=publicDnsName,privateIpAddress
    sshport.default=22
    sshport.selector=tags/ssh_config_Port
    instanceId.selector=instanceId
    nodename.selector=tags/Name,instanceId
    osArch.selector=architecture
    osFamily.default=unix
    osFamily.selector=platform
    osName.default=Linux
    osName.selector=platform
    privateDnsName.selector=privateDnsName
    privateIpAddress.selector=privateIpAddress
    state.selector=state.name
    tag.pending.selector=state.name=pending
    tag.running.selector=state.name=running
    tag.shutting-down.selector=state.name=shutting-down
    tag.stopped.selector=state.name=stopped
    tag.stopping.selector=state.name=stopping
    tag.terminated.selector=state.name=terminated
    tags.default=ec2
    tags.selector=tags/Rundeck-Tags
    username.default=ec2-user
    username.selector=tags/Rundeck-User

Configuring the Mapping
-----------------------

You can configure your source to start with the above default mapping with the 
`useDefaultMapping` property.

You can then selectively change it either by setting the `mappingParams` or 
pointing to a new properties file with `mappingFile`.

For example, you can put this in the `mappingParams` field in the GUI to change 
the default tags for your nodes, remove the "stopping" tag selector, and add a
new "ami_id" selector:

    tags.default=mytag, mytag2;tag.stopping.selector=;ami_id.selector=imageId

Mapping format
---------------

The mapping consists of defining either a selector or a default for
the desired Node fields.  The "nodename" field is required, and will 
automatically be set to the instance ID if no other value is defined.

For purposes of the mapping definition, a `field selector` is either:

* An EC2 fieldname, or dot-separated field names
* "tags/" followed by a Tag name, e.g. "tags/My Tag"
* "tags/*" for use by the `attributes.selector` mapping

Selectors use the Apache [BeanUtils](http://commons.apache.org/beanutils/) to extract a property value from the AWS API
[Instance class](http://docs.amazonwebservices.com/AWSJavaSDK/latest/javadoc/com/amazonaws/services/ec2/model/Instance.html).
This means you can use dot-separated fieldnames to traverse the object graph.
E.g. "state.name" to specify the "name" field of the State property of the Instance.

format:

    # define a selector for "property":
    <attribute>.selector=<field selector>
    # define a default value for "property":
    <attribute>.default=<default value>
    # Special attributes selector to map all Tags to attributes
    attributes.selector=tags/*
    # The value for the tags selector will be treated as a comma-separated list of strings
    tags.selector=<field selector>
    # the default tags list
    tags.default=a,b,c
    # Define a single tag <name> which will be set if and only if the selector result is not empty
    tag.<name>.selector=<field selector>
    # Define a single tag <name> which will be set if the selector result equals the <value>
    tag.<name>.selector=<field selector>=<value>

Note, a ".selector" value can have multiple selectors defined, separated by commas,
and they will be evaluated in order with the first value available being used.  E.g. "nodename.selector=tags/Name,instanceId", which will look for a tag named "Name", otherwise use the instanceId.

You can also use the `<field selector>=<value>` feature to set a tag only if the field selector has a certain value.

### Tags selector

When defining field selector for the `tags` node property, the string value selected (if any) will
be treated as a comma-separated list of strings to use as node tags.  You could, for example, set a custom EC2 Tag on
an instance to contain this list of tags, in this example from the simplemapping.properties file:

    tags.selector=tags/Rundeck-Tags

So creating the "Rundeck-Tags" Tag on the EC2 Instance with a value of "alpha, beta" will result in the node having
those two node tags.

The tags.selector also supports a "merge" ability, so you can merge multiple Instance Tags into the RunDeck tags by separating multiple selectors with a "|" character:

    tags.selector=tags/Environment|tags/Role

### Appending values

A field selector can conjoin multiple values using `+`, and can append literal text like the `_` character for example.

    # conjoin two fields with no separation between the values
    # this will result in "field1field2" 
    <attribute>.selector=<field selector>+<field2 selector>
    
    # conjoin multiple fields with a literal string delimiter
    # this will result in "field1-*-field2"
    <attribute>.selector=<field selector>+"-*-"+<field2 selector>

Use a quoted value to insert a delimiter, with either single or double quotes.

Here is an example to use the "Name" instance tag, and InstanceId, to generate
a unique node name for rundeck:

    nodename.selector=tags/Name+'-'+instanceId


Mapping EC2 Instances to Rundeck Nodes
=================

Rundeck node definitions specify mainly the pertinent data for connecting to and organizing the Nodes.  EC2 Instances have metadata that can be mapped onto the fields used for Rundeck Nodes.

Rundeck nodes have the following metadata fields:

* `nodename` - unique identifier
* `hostname` - IP address/hostname to connect to the node
* `sshport` - The ssh port, if resolved to another port than 22 hostname will be set to ``<hostname>:<sshport>``
* `username` - SSH username to connect to the node
* `description` - textual description
* `osName` - OS name
* `osFamily` - OS family: unix, windows, cygwin.
* `osArch` - OS architecture
* `osVersion` - OS version
* `tags` - set of labels for organization
* `editUrl` - URL to edit the definition of this node object
* `remoteUrl` - URL to edit the definition of this node object using Rundeck-specific integration

In addition, Nodes can have arbitrary attribute values.

EC2 Instance Field Selectors
-----------------

EC2 Instances have a set of metadata that can be mapped to any of the Rundeck node fields, or to Settings or tags for the node.

EC2 fields:

* amiLaunchIndex
* architecture
* clientToken
* imageId
* instanceId
* instanceLifecycle
* instanceType
* kernelId
* keyName
* launchTime
* license
* platform
* privateDnsName
* privateIpAddress
* publicDnsName
* publicIpAddress
* ramdiskId
* rootDeviceName
* rootDeviceType
* spotInstanceRequestId
* state
* stateReason
* stateTransitionReason
* subnetId
* virtualizationType
* vpcId
* `tags/*`

EC2 Instances can also have "Tags" which are key/value pairs attached to the Instance.  A common Tag is "Name" which could be a unique identifier for the Instance, making it a useful mapping to the Node's name field.  Note that EC2 Tags differ from Rundeck Node tags: Rundeck tags are simple string labels and are not key/value pairs.

Authenticating to EC2 Nodes with Rundeck
-----------

Once you get your EC2 Instances listed in Rundeck as Nodes, you may be wondering "Now how do I use this?"

Rundeck uses SSH by default with private key authentication, so in order to connect to your EC2 instances out
of the box you will need to configure Rundeck to use the right private SSH key to connect to your nodes,
which can be done in either of a few ways:

1. Copy your private key to the default location used by Rundeck which is `~/.ssh/id_rsa`
2. Copy your private key elsewhere, and override it on a project level. Change project.properties and set the `project.ssh-keypath` to point to the file.
3. Copy your private key elsewhere, and set the location as an attribute on your nodes (shown below)

To set the ssh keypath attribute on the EC2 Nodes produced by the plugin, you can modify your mapping configuration.

E.g. in the "Mapping Params" field, set:

`Mapping Params: ssh-keypath.default=/path/to/key`

This will set the "ssh-keypath" attribute on your EC2 Nodes, allowing correct private key ssh authentication.

The default mapping also configures a default `username` attribute to be `ec2-user`, but if you want to change the default set:

`Mapping Params: ssh-keypath.default=/path/to/key;username.default=my-username`



Release-Notes.md
Release Notes
=========

1.5.6
-----

Date: 2018-08-15

Changes:

* Support multiple values in aws filter: [#80](https://github.com/rundeck-plugins/rundeck-ec2-nodes-plugin/pull/80)
	* use "FilterName=Value1,Value2;FilterName2=Value1,Value2" etc
* Fixed [#56](https://github.com/rundeck-plugins/rundeck-ec2-nodes-plugin/issues/56) missing clientConfiguration from STS api call

1.5.5
-----

Date: 2017-06-19

Changegs:

* Fix [#67](https://github.com/rundeck-plugins/rundeck-ec2-nodes-plugin/issues/67) Update aws-sdk to 1.11.148

1.5.4
-----

Date: 2017-02-23

Changes:

* Fix [#61](https://github.com/rundeck-plugins/rundeck-ec2-nodes-plugin/issues/61) add synchronousLoad to disable asynchronous behavior

1.5.3
-----

Date: 2016-12-16

Changes:

* fix [#44](https://github.com/rundeck-plugins/rundeck-ec2-nodes-plugin/issues/44) default hostname selector should fallback to privateIpAddress
* fix [#10](https://github.com/rundeck-plugins/rundeck-ec2-nodes-plugin/issues/10) add conjoined selector values using `+` and quoted literals

1.5.2
---

Date: 2016-03-04

Changes:

* fix metadata for plugin file version [#38](https://github.com/rundeck-plugins/rundeck-ec2-nodes-plugin/issues/38)
* add support for assumeRole for IAM profile [#40](https://github.com/rundeck-plugins/rundeck-ec2-nodes-plugin/pull/40)


1.5.1
---

Date: 10/15/2015

Changes:

* add support for setting ssh port [#33](https://github.com/rundeck-plugins/rundeck-ec2-nodes-plugin/pull/33)

1.5
---

Date: 12/10/2014

Changes:

* Fix issue of adding http proxy support [Issue #14](https://github.com/rundeck-plugins/rundeck-ec2-nodes-plugin/issues/14)
* Make AWS access key and secret optional, and use IAM profile instead [#20](https://github.com/rundeck-plugins/rundeck-ec2-nodes-plugin/issues/20)
* Edit link for nodes updated for new AWS console URL scheme
* Use Password field input for AWS SecretKey and Proxy Password
* Fix use of "|" in `tags.selector` mapping [#18](https://github.com/rundeck-plugins/rundeck-ec2-nodes-plugin/issues/18)
* Fix: Errant whitespace in project.properties can cause AWS auth to fail [#13](https://github.com/rundeck-plugins/rundeck-ec2-nodes-plugin/issues/13)

1.4
---

Date: 1/24/2014

Changes:

* Fix log4j problem with Rundeck 1.6
* works with Rundeck 1.6/2.x

1.3
---

Date: 2/22/2013

Changes:

* Update to work with Rundeck 1.5
* Fix incorrect description of Mapping Params config property

1.2
---

Date: 11/16/2011

Changes:

* Added "merge" feature to `tags.selector` mapping

1.1
---

Date: 10/5/2011

Changes:

* Fix issue executing on nodes [Issue #1](https://github.com/gschueler/rundeck-ec2-nodes-plugin/issues/1)

1.0
---

Initial release

Date: 9/9/2011


src/main/java/com/dtolabs/rundeck/plugin/resources/ec2/EC2ResourceModelSourceFactory.java
/*
 * Copyright 2011 DTO Solutions, Inc. (http://dtosolutions.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/*
* EC2ResourceModelSource.java
* 
* User: Greg Schueler <a href="mailto:greg@dtosolutions.com">greg@dtosolutions.com</a>
* Created: 9/1/11 4:27 PM
* 
*/
package com.dtolabs.rundeck.plugin.resources.ec2;

import com.dtolabs.rundeck.core.common.Framework;
import com.dtolabs.rundeck.core.plugins.Plugin;
import com.dtolabs.rundeck.core.plugins.configuration.*;
import com.dtolabs.rundeck.core.resources.ResourceModelSource;
import com.dtolabs.rundeck.core.resources.ResourceModelSourceFactory;
import com.dtolabs.rundeck.plugins.util.DescriptionBuilder;
import com.dtolabs.rundeck.plugins.util.PropertyBuilder;

import java.io.File;
import java.util.*;

/**
 * EC2ResourceModelSourceFactory is the factory that can create a {@link ResourceModelSource} based on a configuration.
 * <p/>
 * The configuration properties are: <ul> <li>endpoint: the AWS endpoint to use, or blank for the default</li>
 * <li>filter: A set of ";" separated query filters ("filter=value") for the AWS EC2 API, see <a
 * href="http://docs.amazonwebservices.com/AWSEC2/latest/APIReference/ApiReference-query-DescribeInstances.html">DescribeInstances</a></li>
 * <li>mappingParams: A set of ";" separated mapping entries</li> <li>runningOnly: if "true", automatically filter the
 * instances by "instance-state-name=running"</li> <li>accessKey: API AccessKey value</li> <li>secretKey: API SecretKey
 * value</li> <li>mappingFile: Path to a java properties-formatted mapping definition file.</li> <li>refreshInterval:
 * Time in seconds used as minimum interval between calls to the AWS API.</li> <li>useDefaultMapping: if "true", base
 * all mapping definitions off the default mapping provided. </li> </ul>
 *
 * @author Greg Schueler <a href="mailto:greg@dtosolutions.com">greg@dtosolutions.com</a>
 */
@Plugin(name = "aws-ec2", service = "ResourceModelSource")
public class EC2ResourceModelSourceFactory implements ResourceModelSourceFactory, Describable {
    public static final String PROVIDER_NAME = "aws-ec2";
    private Framework framework;

    public static final String ENDPOINT = "endpoint";
    public static final String FILTER_PARAMS = "filter";
    public static final String MAPPING_PARAMS = "mappingParams";
    public static final String RUNNING_ONLY = "runningOnly";
    public static final String ACCESS_KEY = "accessKey";
    public static final String SECRET_KEY = "secretKey";
    public static final String ROLE_ARN = "assumeRoleArn";
    public static final String MAPPING_FILE = "mappingFile";
    public static final String REFRESH_INTERVAL = "refreshInterval";
    public static final String SYNCHRONOUS_LOAD = "synchronousLoad";
    public static final String USE_DEFAULT_MAPPING = "useDefaultMapping";
    public static final String HTTP_PROXY_HOST = "httpProxyHost";
    public static final String HTTP_PROXY_PORT = "httpProxyPort";
    public static final String HTTP_PROXY_USER = "httpProxyUser";
    public static final String HTTP_PROXY_PASS = "httpProxyPass";

    public EC2ResourceModelSourceFactory(final Framework framework) {
        this.framework = framework;
    }

    public ResourceModelSource createResourceModelSource(final Properties properties) throws ConfigurationException {
        final EC2ResourceModelSource ec2ResourceModelSource = new EC2ResourceModelSource(properties);
        ec2ResourceModelSource.validate();
        return ec2ResourceModelSource;
    }

    static Description DESC = DescriptionBuilder.builder()
            .name(PROVIDER_NAME)
            .title("AWS EC2 Resources")
            .description("Produces nodes from AWS EC2")

            .property(PropertyUtil.string(ACCESS_KEY, "Access Key", "AWS Access Key", false, null))
            .property(
                    PropertyUtil.string(
                            SECRET_KEY,
                            "Secret Key",
                            "AWS Secret Key, required if Access Key is used. If not used, then the IAM profile will be used",
                            false,
                            null,
                            null,
                            null,
                            Collections.singletonMap("displayType", (Object) StringRenderingConstants.DisplayType.PASSWORD)
                    )
            )
            .property(
                    PropertyUtil.string(
                            ROLE_ARN,
                            "Assume Role ARN",
                            "IAM Role ARN to assume, if using IAM Profile only.\n\nSee [IAM Roles](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles.html).",
                            false,
                            null,
                            null,
                            null
                    )
            )
                                                .property(PropertyUtil.bool(
                                                        SYNCHRONOUS_LOAD,
                                                        "Synchronous Loading",
                                                        "Do not use internal async loading behavior.\n\n" +
                                                        "Note: Rundeck 2.6.3+ uses an asynchronous nodes cache by " +
                                                        "default. You should enable this if you are using the " +
                                                        "rundeck nodes cache.",
                                                        false,
                                                        "true"
                                                ))
            .property(PropertyUtil.integer(REFRESH_INTERVAL, "Async Refresh Interval",
                    "Unless using Synchronous Loading, minimum time in seconds between API requests to AWS (default is 30)", false, "30"))
                                                .property(PropertyUtil.string(
                                                        FILTER_PARAMS,
                                                        "Filter Params",
                                                        "AWS EC2 filters, in the form `Filter=Value`.\n\nYou can "
                                                        + "specify multiple filters by separating them with `;`, and "
                                                        + "you can specify multiple values by separating them with `,"
                                                        + "`.  See [AWS DescribeInstances](https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_DescribeInstances.html) for more information about filters.\n\n"
                                                        + "Example: `tag:MyTag=Some Tag Value;instance-type=m1.small,m1.large`",
                                                        false,
                                                        null
                                                ))
            .property(PropertyUtil.string(ENDPOINT, "Endpoint", "AWS EC2 Endpoint, or blank for default", false, null))
            .property(PropertyUtil.string(HTTP_PROXY_HOST, "HTTP Proxy Host", "HTTP Proxy Host Name, or blank for default", false, null))
            .property(PropertyUtil.integer(HTTP_PROXY_PORT, "HTTP Proxy Port", "HTTP Proxy Port, or blank for 80", false, "80"))
            .property(PropertyUtil.string(HTTP_PROXY_USER, "HTTP Proxy User", "HTTP Proxy User Name, or blank for default", false, null))
            .property(
                    PropertyUtil.string(
                            HTTP_PROXY_PASS,
                            "HTTP Proxy Password",
                            "HTTP Proxy Password, or blank for default",
                            false,
                            null,
                            null,
                            null,
                            Collections.singletonMap("displayType", (Object) StringRenderingConstants.DisplayType.PASSWORD)
                    )
            )
            .property(PropertyUtil.string(MAPPING_PARAMS, "Mapping Params",
                    "Property mapping definitions. Specify multiple mappings in the form " +
                            "\"attributeName.selector=selector\" or \"attributeName.default=value\", " +
                            "separated by \";\"",
                    false, null))
            .property(PropertyUtil.string(MAPPING_FILE, "Mapping File", "Property mapping File", false, null,
                    new PropertyValidator() {
                        public boolean isValid(final String s) throws ValidationException {
                            if (!new File(s).isFile()) {
                                throw new ValidationException("File does not exist: " + s);
                            }
                            return true;
                        }
                    }))
            .property(PropertyUtil.bool(USE_DEFAULT_MAPPING, "Use Default Mapping",
                    "Start with default mapping definition. (Defaults will automatically be used if no others are " +
                            "defined.)",
                    false, "true"))
            .property(PropertyUtil.bool(RUNNING_ONLY, "Only Running Instances",
                    "Include Running state instances only. If false, all instances will be returned that match your " +
                            "filters.",
                    false, "true"))

            .build();

    public Description getDescription() {
        return DESC;
    }
}
